### PR TITLE
Fix race between setting tick height and calculating accounts hash

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Freeze in ReplayStage only checks the tick height: https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L1339-L1348), so it's possible that the the accounts diff hash is calculated after freeze, but before the block hash sysvar is updated. Like this:
```
let current_tick_height = self.tick_height.fetch_add(1, Relaxed) as u64;
<<< Freeze happens here>>>
if self.is_block_boundary(current_tick_height + 1) {
            w_blockhash_queue.register_hash(hash, &self.fee_calculator);
            if self.fix_recent_blockhashes_sysvar_delay() {
                self.update_recent_blockhashes_locked(&w_blockhash_queue);
            }
        }
```

#### Summary of Changes
Move blockhash registration before tick registration.

TODO: Testing and double checking subtleties in old related PR's that I may have forgotten about:

1) https://github.com/solana-labs/solana/pull/6013
2) https://github.com/solana-labs/solana/pull/6799

Fixes #

### CODE REVIEW

**Comment by ryoqun** [comment]
Thanks for fixing this!

(Also, I wonder why this problem happens so often at our internal testing than at any of live clusters? Naturally, this fix explain that?)

**Comment by sakridge** [comment]
Nice, this seems to make sense. Is there an assert we can add on doing a store on a frozen bank that makes sense to catch this kind of thing earlier?

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -2268,13 +2268,13 @@ impl Bank {
...
         let mut w_blockhash_queue = self.blockhash_queue.write().unwrap();
-        let current_tick_height = self.tick_height.fetch_add(1, Relaxed) as u64;
-        if self.is_block_boundary(current_tick_height + 1) {
+        if self.is_block_boundary(self.tick_height.load(Relaxed) + 1) {
             w_blockhash_queue.register_hash(hash, &self.fee_calculator);
             if self.fix_recent_blockhashes_sysvar_delay() {
                 self.update_recent_blockhashes_locked(&w_blockhash_queue);
             }
         }
+        let current_tick_height = self.tick_height.fetch_add(1, Relaxed) as u64;
```
Comment (sakridge) [inline runtime/src/bank.rs:2277]: Maybe a comment about why this needs to be last in the function?

**Comment by carllin** [comment]
> Nice, this seems to make sense. Is there an assert we can add on doing a store on a frozen bank that makes sense to catch this kind of thing earlier?

Added to AccountsDb to catch frozen stores, what do you think?

**Comment by carllin** [comment]
Been running this on a validator against mainnet, and also a 5 node cluster with 5k tps, no issues thus far.

**Comment by sakridge** [comment]
> > Nice, this seems to make sense. Is there an assert we can add on doing a store on a frozen bank that makes sense to catch this kind of thing earlier?
> 
> Added check: https://github.com/solana-labs/solana/pull/14101/files#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR2653 to AccountsDb to catch frozen stores, what do you think?

How about just similar assert in `Bank::store_account` to this:
```
assert!(
    !self.is_frozen(),
    "commit_transactions() working on a frozen bank!"
);
```

**Comment by carllin** [comment]
@sakridge, yeah I considered something similar, but the issue is `is_frozen() == true` is set only after `*hash = self.hash_internal_state();`.

This means  when the `let accounts_delta_hash = self.rc.accounts.bank_hash_info_at(self.slot());` inside `hash_internal_state()` is being computed, `is_frozen() == false`. Thus if the `store()` happens after the `accounts_delta_hash` computation and before the rest of the `hash_internal_state()` function finishes, the assert wouldn't trigger.

**Comment by sakridge** [comment]
> @sakridge, yeah I considered something similar, but the issue is `is_frozen() == true` is set only after `*hash = self.hash_internal_state();`.
> 
> This means when the `let accounts_delta_hash = self.rc.accounts.bank_hash_info_at(self.slot());` inside `hash_internal_state()` is being computed, `is_frozen() == false`. Thus if the `store()` happens after the `accounts_delta_hash` computation and before the rest of the `hash_internal_state()` function finishes, the assert wouldn't trigger.

I see. `is_frozen: AtomicBool` in the bank?

**Comment by carllin** [comment]
> > @sakridge, yeah I considered something similar, but the issue is `is_frozen() == true` is set only after `*hash = self.hash_internal_state();`.
> > This means when the `let accounts_delta_hash = self.rc.accounts.bank_hash_info_at(self.slot());` inside `hash_internal_state()` is being computed, `is_frozen() == false`. Thus if the `store()` happens after the `accounts_delta_hash` computation and before the rest of the `hash_internal_state()` function finishes, the assert wouldn't trigger.
> 
> I see. `is_frozen: AtomicBool` in the bank?

Hehe nope, it checks the blockhash like so:

```
pub fn is_frozen(&self) -> bool {
        *self.hash.read().unwrap() != Hash::default()
    }
```

I added it as a method on AccountsDb so it would catch it regardless of the storage entry point (i.e. bank.store() vs bank.accounts.store_accounts() in bank.commit_transactions())

**Comment by sakridge** [comment]
> > > @sakridge, yeah I considered something similar, but the issue is `is_frozen() == true` is set only after `*hash = self.hash_internal_state();`.
> > > This means when the `let accounts_delta_hash = self.rc.accounts.bank_hash_info_at(self.slot());` inside `hash_internal_state()` is being computed, `is_frozen() == false`. Thus if the `store()` happens after the `accounts_delta_hash` computation and before the rest of the `hash_internal_state()` function finishes, the assert wouldn't trigger.
> > 
> > 
> > I see. `is_frozen: AtomicBool` in the bank?
> 
> Hehe nope, it checks the blockhash like so:
> 
> ```
> pub fn is_frozen(&self) -> bool {
>         *self.hash.read().unwrap() != Hash::default()
>     }
> ```
> 
> I added it as a method on AccountsDb so it would catch it regardless of the storage entry point (i.e. bank.store() vs bank.accounts.store_accounts() in bank.commit_transactions())

Yea, I get the trade-off. But it is propagating this notion of freezing down to an additional layer and also adding a whole other structure to take care of and prune vs. adding it in those two places which we do a store. Also taking additional locks for duplicate functionality.

I think it might work like this:

```
Bank {
  ..
  is_frozen: AtomicBool,
}
fn is_frozen(&self) -> bool { self.is_frozen.load(..) }
fn is_frozen_and_hashed(&self) -> bool { self.is_frozen() && self.hash.read().unwrap() != default() }
fn freeze(&self) {
    let mut hash = self.hash.write().unwrap();
    store_xyz();
    self.is_frozen.store(true, ..);
    *hash = self.hash_internal_state();
}
```

**Comment by carllin** [comment]
@sakridge I followed your suggestion except I renamed `is_frozen -> freeze_started` and `is_frozen_and_hashed -> is_frozen` (i.e. is_frozen remains unchanged). This is because:

1) In your examples, `self.is_frozen() && self.hash.read().unwrap() != default()` is equivalent to `self.hash.read().unwrap() != default()`, because the  `is_frozen` AtomicBool must be set if `self.hash.read().unwrap() != default()`.
2) Defaulting to the original `is_frozen` seems safer than making the rename change because I am not sure if there are implicit dependencies on blockhash in a few places.

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -3770,6 +3790,7 @@ impl Bank {
     }
 
     pub fn store_account(&self, pubkey: &Pubkey, account: &Account) {
+        assert!(self.freeze_started());
```
Comment (carllin) [inline runtime/src/bank.rs:3793]: Second place we store, added check

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -3055,8 +3075,8 @@ impl Bank {
         signature_count: u64,
     ) -> TransactionResults {
         assert!(
-            !self.is_frozen(),
-            "commit_transactions() working on a frozen bank!"
+            !self.freeze_started(),
```
Comment (carllin) [inline runtime/src/bank.rs:3078]: First place we store, added check

File: runtime/src/bank.rs [hunk 4]
```diff
@@ -1145,6 +1148,7 @@ impl Bank {
             transaction_log_collector: new(),
             feature_set: new(),
             drop_callback: RwLock::new(OptionalDropCallback(None)),
+            freeze_started: AtomicBool::new(fields.hash != Hash::default()),
```
Comment (carllin) [inline runtime/src/bank.rs:1151]: @ryoqun I think this is a safe change to the ABI, and safe functionality-wise

Comment (ryoqun) [reply runtime/src/bank.rs:1151]: yeah, I think this is ok.

**Comment by sakridge** [comment]
> 
> 
> @sakridge I followed your suggestion: [605f14a](https://github.com/solana-labs/solana/commit/605f14acc86f8ac0ff13e4e7d99756a1d09bb8c4) except I introduced a new ` freeze_started` flag (serves the same purpose as the `is_frozen` AtomicBool in your example), and `is_frozen_and_hashed -> is_frozen` (i.e. `is_frozen` remains unchanged). This is because:
> 
>     1. In your examples, `self.is_frozen() && self.hash.read().unwrap() != default()` is equivalent to `self.hash.read().unwrap() != default()`, because the  `is_frozen` AtomicBool must be set if `self.hash.read().unwrap() != default()`, so I think just having the name `is_frozen()` is sufficient
> 
>     2. Defaulting to the original `is_frozen` seems safer than making the rename change because I am not sure if there are implicit dependencies on blockhash in a few places.

Yes, perfect. Those names are better. Thanks. Key point was just that an atomic load is much cheaper than rwlock+ btree operation.

**Pull request merged**